core/src/main/java/org/ldaptive/url/Url.java

```diff
@@ -0,0 +1,443 @@
+/* See LICENSE for licensing and NOTICE for copyright. */
+package org.ldaptive.url;
+
+import java.util.Arrays;
+import org.ldaptive.LdapUtils;
+import org.ldaptive.SearchScope;
+import org.ldaptive.dn.Dn;
+import org.ldaptive.filter.Filter;
+import org.ldaptive.filter.PresenceFilter;
+
+/**
+ * Class to represent an LDAP URL. See RFC 4516. Expects URLs of the form
+ * scheme://hostname:port/baseDn?attrs?scope?filter. This implementation does not support URL extensions.
+ *
+ * @author  Middleware Services
+ */
+public final class Url
+{
+
+  /** Default LDAP port, value is {@value}. */
+  static final int DEFAULT_LDAP_PORT = 389;
+
+  /** Default LDAPS port, value is {@value}. */
+  static final int DEFAULT_LDAPS_PORT = 636;
+
+  /** Default LDAPI port, value is {@value}. */
+  static final int DEFAULT_LDAPI_PORT = 0;
+
+  /** Default base DN, value is {@value}. */
+  static final String DEFAULT_BASE_DN = "";
+
+  /** Default parsed base DN. */
+  static final Dn DEFAULT_PARSED_BASE_DN = new Dn();
+
+  /** Default attribute filter value is '(objectClass=*)'. */
+  static final String DEFAULT_FILTER = "(objectClass=*)";
+
+  /** Default parsed attribute filter. */
+  static final Filter DEFAULT_PARSED_FILTER = new PresenceFilter("objectClass");
+
+  /** Default scope, value is {@link SearchScope#OBJECT}. */
+  static final SearchScope DEFAULT_SCOPE = SearchScope.OBJECT;
+
+  /** Default return attributes, value is all user attributes. */
+  static final String[] DEFAULT_ATTRIBUTES = new String[0];
+
+  /** hash code seed. */
+  private static final int HASH_CODE_SEED = 11003;
+
+  /** Scheme of the ldap url. */
+  private final String scheme;
+
+  /** Hostname of the ldap url. */
+  private final String hostname;
+
+  /** Port of the ldap url. */
+  private final int port;
+
+  /** Base DN of the ldap url. */
+  private final String baseDn;
+
+  /** Parsed base DN of the ldap url. */
+  private final Dn parsedBaseDn;
```

Comment:
You are representing the DN twice in the underlying data model. While I can appreciate that storing both the parse and unparsed representations may be helpful, you should not expose this detail externally. Pick one for your constructors and derive the other if necessary.

Reply:
How the DN is parsed is actually an implementation detail for the `UrlParser`, which is why it's provided and not derived. I really wanted a single parameter, like a supplier that produces a tuple. But that produced more code than it was worth.

Reply:
I feel pretty strongly that this is a sign of a need for refinement. Taking what you said as fact, then I would augment the `UrlParser` API to address DN parsing in addition to LDAP URL parsing. Then you could use composition in your method calls to avoid duplication. I'm squinting to see the solution, but my instinct says that's one way to keep concerns as you have them and avoid this duplication.

Reply:
It sounds like you'd prefer the `Url` interface to change from:
```
Url parse(String url);
````
to
```
Url parse(String url, DnParser dnParser, FilterFunction filterFunction);
```

Reply:
I can also remove the parsed fields and leave parsing as an exercise for the `UrlFormatter` implementations. That's probably the cleanest solution.

Reply:
I think your instinct is right, but I want to avoid over-engineering here. You don't have a use case for the parsed DN given the `UrlFormatter` you provide with the library, `MinimalUrlFormatter`. If you feel strongly that working on `Dn` object may be useful at some point, you'd be better off generifying the interface to accept an arbitrary type for the DN, and then providing `MinimalUrlFormatter<String>` with a string representation of the DN. The you've given rope to provide a `CustomUrlFormatter<Dn>` in the future. All that said, is it worth it?

Reply:
it is better to have strong types, String as dns is not valid, but as Dn then you know is valid

Reply:
I ended up internalizing the parsing of both the `parsedDn` and the `parsedFilter`. So `Url` can expose both the original version and the parsed version. It's not the cleanest implementation, but I think it's the most flexible. @serac have a look at the latest.

Reply:
This addresses my biggest concern, so we can move forward.